### ISSUE TITLE
Feature Request: Provide Content-Security-Policy 'nonce' attribute

### ISSUE THREAD

**Issue by csmk** [h2o/h2o#1218]
Hi,

I don't know if this is already possible with mruby, but if not, I think it's a nice security feature to have. There is a [WIP implementation](https://github.com/mholt/caddy/issues/1302) for Caddy server in which I think the H2O devs could use for additional information.

If this is already possible, could someone give some direction on how to proceed? Thanks.

**Comment by kazuho** [comment]
Thank you for opening the issue.

My understanding is that for a dynamically generated content (e.g. content generated by FastCGI), your application can generate the response header and the HTML that uses the correct nonce.

Therefore, I'd assume that you are looking for an output filter (much like the on-the-fly compression handler) that rewrites HTML sent by _any_ handler (`file.dir`, `proxy.reverse.url`) at the same timing setting the nonce value in the response header. Am I correct?

If that is the case, we do not have such thing at the moment, but I am sure the use-case will be covered by https://github.com/h2o/h2o/pull/1217.

**Comment by csmk** [comment]
Hi, Kazuho-san.

Thank you for your time on this matter. Yes, you're correct, that exactly what I'm looking for. Well, I'm happy to see that there is already a WIP, thanks to @i110.

So even after #1217 is done, do you think it will be easy to implement such use case? Or I'll need to write complex functions? Like to generate the nonces.
I'm still learning on how to use mruby, so forgive my ignorance. :blush:

**Comment by kazuho** [comment]
@csmk Thank you for the clarification. I am sorry to ask this once again, but could you please elaborate exactly how you are interested in using the feature?

Generally speaking only the application knows which script tag is legitimate  and which is not. Therefore, it is not H2O but the application that should emit tags with the `nonce` attributes. How would you expect H2O to distinguish between a legitimate script tag (that would have the `nonce` attribute injected) and others? And how would that be easier to handle than emitting both the response header that contains the nonce and the response with the nonce from the application?

**Comment by csmk** [comment]
I'm sorry for not been clear. Well, I was thinking much like the Caddy implementation, in which the application actually do nothing. H2O will just recognize some special tag and rewrite it with the generated `nonce`.

Adding a header, something like this:
```
Content-Security-Policy "default-src 'self'; script-src 'self' {H2OnonceTag}; style-src 'self' {H2OnonceTag};"
```

And then rewriting the body filed.
```
<style nonce="{H2OnonceTag}"> ... </style>
```

**Comment by kazuho** [comment]
@csmk Thank you for explaining.

The approach makes sense for Caddy since it is the one that builds the template. Caddy will convert `{H2OnonceTag}` found in the template, but not the ones that found in the strings that are _injected_ by using the template.

Consider the following case. `{nonce}` will be replaced to the correct value, but `{nonce}` that might appear in `user_comment` will not get replaced.

```
<style nonce="{nonce}">...</style>
<div>{user_comment}</div>
```

This is how the mitigation works.

OTOH, H2O does not build the template. At the moment H2O receives the response from the application server, it is impossible to distinguish between a legitimate `{nonce}` that should be replaced and a `{nonce}` injected by an attacker that would appear within `user_comment`.

To recap, what you should do the rewrites when you process the template. And I assume that you can emit a response header that has the corresponding nonce when you process the template. Therefore, the responsibility of building and applying the nonce belongs to the application, not H2O.

**Comment by csmk** [comment]
I see. Thank your for the detailed explanation. Now I understand the security implications of my initial approach.

I'll look further on how to implement in the application side.

Cheers!
